pyggi/base/program.py

```diff
@@ -274,45 +272,29 @@ def exec_cmd(self, cmd, timeout=15):
             start = time.time()
             stdout, stderr = sprocess.communicate(timeout=timeout)
             end = time.time()
-            result = self.__class__.Result(status_code=StatusCode.NORMAL,
-                            elapsed_time=(end - start),
-                            stdout=stdout.decode("ascii"),
-                            stderr=stderr.decode("ascii"))
+            return (sprocess.returncode, stdout.decode("ascii"), stderr.decode("ascii"), end-start)
         except subprocess.TimeoutExpired:
             sprocess.kill()
-            result = self.__class__.Result(status_code=StatusCode.TIME_OUT,
-                            elapsed_time=None,
-                            stdout=None,
-                            stderr=None)
-        os.chdir(cwd)
-        return result
-
-    def run(self, timeout=15):
-        """
-        Run the test script of the temporary variant
+            return (None, None, None, None)
+        finally:
+            os.chdir(cwd)
 
-        :param float timeout: The time limit of test run (unit: seconds)
-        :return: The fitness value of the patch
-        """
-        return self.exec_cmd(self.test_command, timeout)
-
-    def compute_fitness(self, elapsed_time, stdout, stderr):
+    def compute_fitness(self, result, return_code, stdout, stderr, elapsed_time):
         try:
-            return float(stdout.strip())
+            result.fitness = float(stdout.strip())
         except:
-            raise ParseError
+            result.status = 'PARSE_ERROR'
 
     def evaluate_patch(self, patch, timeout=15):
         # apply + run
         self.apply(patch)
-        result = self.run(timeout)
-        if result.status_code == StatusCode.TIME_OUT:
-            return StatusCode.TIME_OUT, None
-        try:
-            fitness = self.compute_fitness(result.elapsed_time, result.stdout, result.stderr)
-        except ParseError:
-            return StatusCode.PARSE_ERROR, None
-        return StatusCode.NORMAL, fitness
+        return_code, stdout, stderr, elapsed_time = self.exec_cmd(self.test_command, timeout)
+        if return_code is None: # timeout
+            return RunResult('TIMEOUT')
+        else:
+            result = RunResult('SUCCESS', None)
+            self.compute_fitness(result, return_code, stdout, stderr, elapsed_time)
```

Comment:
for safety's sake, we may add an assertion after this statement

```python
assert not (result.status == 'SUCCESS' and result.fitness is None)
```

Reply:
I would rather not make any assumption over the fitness of successful runs, but I really can't think of any use case that would require having a `None` one.
Assertion added.

Reply:
Alternatively, we could also make run statuses `PARSE_ERROR` by default (instead of `SUCCESS`) so that runs need to explicitly be set successful in `compute_fitness()`.

Reply:
I think `SUCCESS` is better 🙂.